### PR TITLE
Making Group and Summary regular components, only node passed to components

### PR DESCRIPTION
**Pull request by olemartinorg**
## Description
Yet another installment after #976 made `GenericComponent` possible render with just the `node` object as input. This time:

- All component classes now only receive `node` from `GenericComponent`, meaning this PR takes this concept one step further. Components no longer should require every component property as input when rendering themselves. A lot of tests had to be rewritten to make this work.
- `Group` and `Summary` has been promoted (or demoted?) to regular components, which render via `GenericComponent`. No more need for a super-function that decides if a component can be rendered using `GenericComponent` or directly rendered (any component can now ask to be directly rendered).
- No need for `SummaryComponentSwitch` anymore, as every component type can now implement how it is to be rendered in a Summary instead of the Summary component having to know about different component types.

## Related Issue(s)
- #976 
- #377

## Verification/QA

- Manual functionality testing
  - [x] I have tested these changes manually
  - [ ] Creator of the original issue (or service owner) has been contacted for manual testing (or will be contacted when released in alpha)
  - [ ] No testing done/necessary
- Automated tests
  - [ ] Unit test(s) have been added/updated
  - [ ] Cypress E2E test(s) have been added/updated
  - [x] No automatic tests are needed here (no functional changes/additions)
  - [ ] I want someone to help me make some tests
- UU/WCAG ([follow these guidelines](https://aksel.nav.no/god-praksis/artikler/utvikling?tema=universell-utforming) until we have our own)
  - [ ] I have tested with a screen reader/keyboard navigation/automated wcag validator
  - [x] No testing done/necessary (no DOM/visual changes)
  - [ ] I want someone to help me perform accessibility testing
- User documentation @ [altinn-studio-docs](https://github.com/Altinn/altinn-studio-docs)
  - [ ] Has been added/updated
  
  - [x] No functionality has been changed/added, so no documentation is needed
  - [ ] I will do that later/have created an issue
  
- Changes/additions to component properties
  - [ ] Changes are reflected in both `src/layout/layout.d.ts` and `layout.schema.v1.json`, and these are all backwards-compatible
  - [x] No changes made
- Support in Altinn Studio
  - [ ] Issue(s) created for support in Studio
  
  - [x] This change/feature does not require any changes to Altinn Studio
- Sprint board
  - [x] The original issue (or this PR itself) has been added to the Team Apps project and to the current sprint board
  - [ ] I don't have permissions to do that, please help me out
- Labels
  - [x] I have added a `kind/*` label to this PR for proper release notes grouping
  - [ ] I don't have permissions to add labels, please help me out

### CODE REVIEW

File: src/features/form/containers/Form.tsx [hunk 1]
```diff
@@ -4,67 +4,12 @@ import Grid from '@material-ui/core/Grid';
...
-import { PanelGroupContainer } from 'src/layout/Panel/PanelGroupContainer';
 import { ReadyForPrint } from 'src/shared/components/ReadyForPrint';
 import { extractBottomButtons, hasRequiredFields } from 'src/utils/formLayout';
 import { useExprContext } from 'src/utils/layout/ExprContext';
 import { getFormHasErrors, missingFieldsInLayoutValidations } from 'src/utils/validation/validation';
-import type { ComponentExceptGroupAndSummary } from 'src/layout/layout';
-import type { LayoutNode } from 'src/utils/layout/hierarchy';
-import type { LayoutNodeFromType } from 'src/utils/layout/hierarchy.types';
-
-export function renderLayoutNode(node: LayoutNode) {
```
Comment (framitdavid) [inline src/features/form/containers/Form.tsx:21]: Nice clean up! 🚀

File: src/layout/AttachmentList/AttachmentListComponent.tsx [hunk 2]
```diff
@@ -41,7 +41,7 @@ export function AttachmentListComponent(props: IAttachmentListProps) {
         className='attachmentList-title'
         component='span'
       >
-        {props.text || ''}
+        {text || ''}
```
Comment (framitdavid) [inline src/layout/AttachmentList/AttachmentListComponent.tsx:44]: I think we don't need to fallback to an empty string here, because null or undefined should not be rendered. 🤔 
```suggestion
        {text}
```

File: src/layout/Checkboxes/CheckboxesContainerComponent.tsx [hunk 3]
```diff
@@ -80,21 +80,15 @@ const defaultOptions: IOption[] = [];
 const defaultSelectedOptions: string[] = [];
 
 export const CheckboxContainerComponent = ({
-  id,
```
Comment (framitdavid) [inline src/layout/Checkboxes/CheckboxesContainerComponent.tsx:83]: Well done with the props, looks much better to group them within the node prop. 👏

File: src/layout/FileUploadWithTag/EditWindowComponent.tsx [hunk 4]
```diff
@@ -152,7 +153,7 @@ export function EditWindowComponent(props: EditWindowProps): JSX.Element {
               disabled={saveIsDisabled}
               className={classNames(classes.select, 'custom-select a-custom-select', {
                 'validation-error': props.attachmentValidations.filter((i) => i.id === props.attachment.id).length > 0,
-                'disabled !important': props.attachment.updating ? true : props.readOnly,
+                'disabled !important': props.attachment.updating ? true : readOnly,
```
Comment (framitdavid) [inline src/layout/FileUploadWithTag/EditWindowComponent.tsx:156]: I think this expression could be simplified to simething like the suggestion?
```suggestion
                'disabled !important': props.attachment.updating || readOnly,
```

_Looks like I just cleaned my glasses. Since I look at small things like this..._

File: src/layout/GenericComponent.tsx [hunk 5]
```diff
@@ -256,10 +257,8 @@ export function GenericComponent<Type extends ComponentExceptGroupAndSummary = C
 
   const componentProps = {
     ...fixedComponentProps,
-    // TODO: Pass on the node object instead of all the properties in it. This could work fairly simply for most
```
Comment (framitdavid) [inline src/layout/GenericComponent.tsx:259]: 👏

File: src/layout/Group/index.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        overrides={overrides}
+      />
+    );
+  }
+
+  renderSummaryBoilerplate(): boolean {
+    return false;
+  }
+
+  useDisplayData(): string {
```
Comment (framitdavid) [inline src/layout/Group/index.tsx:40]: Not related to this PR, but should this method have been renamed to getDisplayData? Within the React ecosystem, I associate the prefix "use" with a react-hook that uses other hooks. By reading the code within this class it looks like a normal function.

But again, maybe some other classes that are also inherent in the FormComponent actually use hooks within the useDisplayData method. 

No change is required - these were just some thoughts while reading the code.

**Review by framitdavid — APPROVED** [review]
As always a lot of great work! 🚀

File: src/components/summary/SummaryComponent.tsx [hunk 7]
```diff
@@ -132,13 +137,27 @@ export function SummaryComponent({ summaryNode, overrides }: ISummaryComponent)
...
         })}
       >
-        <SummaryComponentSwitch
-          summaryNode={summaryNode}
-          targetNode={targetNode}
-          change={change}
-          label={label}
-          overrides={overrides}
-        />
+        {RenderSummary && component instanceof FormComponent ? (
```
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/components/summary/SummaryComponent.tsx:135]: Yup, but typescript doesn't agree with that.. 🤷 It complains because it didn't infer that component must be a FormComponent if RenderSummary is set, so the call to `component.renderSummaryBoilerplate()` below fails without that check.

File: src/components/summary/SummaryComponent.tsx [hunk 8]
```diff
@@ -132,13 +137,27 @@ export function SummaryComponent({ summaryNode, overrides }: ISummaryComponent)
...
+          <>
+            {component.renderSummaryBoilerplate() ? (
+              <SummaryBoilerplate
+                {...change}
```
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/components/summary/SummaryComponent.tsx]: Hah, yeah, not just that - but I messed this up so the feature it provides was broken! 😆 (Meaning, the "change" button disappeared when rendering a summary for a `Group`). Fixed that just before you wrote your review.

File: src/components/summary/SummaryComponent.tsx [hunk 9]
```diff
@@ -132,13 +137,27 @@ export function SummaryComponent({ summaryNode, overrides }: ISummaryComponent)
...
+                label={label}
+                summaryNode={summaryNode}
+                targetNode={targetNode}
+                overrides={overrides}
+              />
+            ) : null}
+            <RenderSummary
+              {...change}
+              summaryNode={summaryNode}
+              targetNode={targetNode as any}
```
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/components/summary/SummaryComponent.tsx:151]: Whoops, hoped nobody saw that thing.. 🙄 At this point, typescript thinks both `RenderSummary` and `targetNode` is a union-type of all possible component types, but our typings specify that `RenderSummary` takes in a `targetNode` of its own type. Typescript can't guarantee that here, but it's self-evident from the code that we're not calling the render function for _some component with an entirely _unrelated_ `targetNode`.

In other words, the component utility class and `node` object are not sufficiently bound together. I have some thoughts on how we could do that, but I haven't figured that one out quite yet.

File: src/layout/Group/index.tsx (hunk 6 above)
Comment (olemartinorg) [reply src/layout/Group/index.tsx:40]: > But again, maybe some other classes that are also inherent in the FormComponent actually use hooks within the useDisplayData method.

Spot on! 😉 "most" components fetch their form data in this step, which many of them get from redux, meaning these methods are usually using hooks to fetch the data they need. This is why I had to disable the hook rules for these files, since the eslint rules thinks these files contain class components (which they don't, but it kinda looks like it). I would agree that it's a bit funky, though, and I think we can move the target a bit further and find a way to work around this while keeping our code somewhat nice.

I added a `getFormData()` function to `LayoutNode`, since formdata is an input when resolving the node tree anyway, so `LayoutNode` can find data in data bindings. That still doesn't work for a `list` binding in, for example, the `FileUpload` component - so that has to figure out the list binding on its own. There's still a way to go here, in other words!

File: src/components/summary/SummaryComponent.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (framitdavid) [reply src/components/summary/SummaryComponent.tsx:151]: Great! We do not need to fix this within this PR. Hopefully, we could talk about it and figure out how to fix it to avoid this next time. 👍

File: src/layout/Group/index.tsx (hunk 6 above)
Comment (framitdavid) [reply src/layout/Group/index.tsx:40]: Thanks for the clarification! Anyway, we are in the right direction to make the codebase more maintainable. Great work! 🚀 For now I'm resolving this comment and we can discuss this later on. Maybe we somewhere in the future can fetch data and handle states differently. :)
